File: constraints/dataAnalysis/stellarMassFunctions_VIPERS_z0_1/massDistanceRelation.pl
#!/usr/bin/env perl
use strict;
use warnings;
my $galacticusPath;
if ( exists($ENV{"GALACTICUS_ROOT_V093"}) ) {
 $galacticusPath  = $ENV{"GALACTICUS_ROOT_V093"};
 $galacticusPath .= "/" unless ( $galacticusPath =~ m/\/$/ );
} else {
 $galacticusPath  = "./";
}
unshift(@INC,$galacticusPath."perl"); 
use PDL;
use PDL::NiceSlice;
use PDL::Fit::Polynomial;
use XML::Simple;
use Data::Dumper;
use Astro::Cosmology;
require Stats::Percentiles;
require GnuPlot::PrettyPlots;
require GnuPlot::LaTeX;

# Determine the relation between stellar mass and limiting distance for the VIPERS stellar mass functions.
# Andrew Benson (04-June-2014)

# Define a working directory.
my $workDirectory = $galacticusPath."constraints/dataAnalysis/stellarMassFunctions_VIPERS_z0_1/";

# Define survey solid angle (computed from mangle polygons).
my $solidAngle = pdl 0.003137; #0.004143678167878;

# Read the stellar mass function.
my $xml = new XML::Simple();
my $stellarMassFunction = $xml->XMLin($galacticusPath."data/observations/massFunctionsStellar/Stellar_Mass_Functions_VIPERS_2013.xml");

# Construct cosmology.
my $cosmologyObserved 
    = Astro::Cosmology->new(
    omega_matter => $stellarMassFunction->{'cosmology'}->{'omegaMatter'    }, 
    omega_lambda => $stellarMassFunction->{'cosmology'}->{'omegaDarkEnergy'}, 
    h0           => $stellarMassFunction->{'cosmology'}->{'hubble'         }
    );

# Estimated sampling rates for each redshift.
my $samplingRate = pdl [ 0.44, 0.44, 0.4 ];

# Create a plot.
my $plot;
my $gnuPlot;
my $plotFile = $workDirectory."massDistanceRelation.pdf";
(my $plotFileEPS = $plotFile) =~ s/\.pdf$/.eps/;
open($gnuPlot,"|gnuplot 1>/dev/null 2>&1");
print $gnuPlot "set terminal epslatex color colortext lw 2 solid 7\n";
print $gnuPlot "set output '".$plotFileEPS."'\n";
print $gnuPlot "set title 'Limiting Distance vs. Stellar Mass for VIPERS Survey'\n";
print $gnuPlot "set xlabel 'Stellar mass; \$M_\\star [M_\\odot]\$'\n";
print $gnuPlot "set ylabel 'Limiting distance; \$D_{\\rm max}\$ [Mpc]'\n";
print $gnuPlot "set lmargin screen 0.15\n";
print $gnuPlot "set rmargin screen 0.95\n";
print $gnuPlot "set bmargin screen 0.15\n";
print $gnuPlot "set tmargin screen 0.95\n";
print $gnuPlot "set key spacing 1.2\n";
print $gnuPlot "set key at screen 0.35,0.7\n";
print $gnuPlot "set key left\n";
print $gnuPlot "set key bottom\n";
print $gnuPlot "set logscale xy\n";
print $gnuPlot "set mxtics 10\n";
print $gnuPlot "set mytics 10\n";
print $gnuPlot "set format x '\$10^{\%L}\$'\n";
print $gnuPlot "set format y '\$10^{\%L}\$'\n";
print $gnuPlot "set xrange [3.0e9:3.0e12]\n";
print $gnuPlot "set yrange [2000:4000]\n";
print $gnuPlot "set pointsize 2.0\n";

# Iterate over redshifts.
my $redshiftBin = -1;
for my $columns ( @{$stellarMassFunction->{'massFunction'}->{'columns'}} ) {
    ++$redshiftBin;

    # Extract mass function.
    my $logarithmicMass = pdl $columns->{'mass'        }->{'datum'};
    my $massFunction    = pdl $columns->{'massFunction'}->{'datum'};
    my $number          = pdl $columns->{'number'      }->{'datum'};
    my $binWidth        = pdl 0.2*ones(nelem($massFunction));
    my $redshiftMinimum = pdl $columns->{'redshiftLow' };
    my $redshiftMaximum = pdl $columns->{'redshiftHigh'};
    my $redshiftLabel   = sprintf("\$%3.1f<z<%3.1f\$",$redshiftMinimum->sclr(),$redshiftMaximum->sclr());

    # Find distance to minimum and maximum redshifts.
    my $distanceMinimum = $cosmologyObserved->comoving_distance($redshiftMinimum);
    my $distanceMaximum = $cosmologyObserved->comoving_distance($redshiftMaximum);

    # Compute the effective volume.
    my $volume = $number/$samplingRate->(($redshiftBin))/$massFunction/$binWidth;
 
    # Find the maximum distance.
    my $distance            = (3.0*$volume/$solidAngle+$distanceMinimum**3)**(1.0/3.0);
    my $logarithmicDistance = log10($distance);

    # Fit a polynomial to the results.
    my $fitTo             = which(($massFunction > 0.0) & ($distance < 0.98*$distanceMaximum) & ($logarithmicMass < 12));
    (my $fit, my $coeffs) = fitpoly1d($logarithmicMass->($fitTo),$logarithmicDistance->($fitTo),2);

    # Generate a fit to the data.
    my $fitMass     = sequence(1000)*9.0/999.0+4.0;
    my $fitDistance = pdl zeroes(nelem($fitMass));
    my $closing     = "";
    print $redshiftMinimum." < z < ".$redshiftMaximum.": ";
    for(my $i=0;$i<nelem($coeffs);++$i) {
	$fitDistance += $coeffs->(($i))*($fitMass**$i);
	print $coeffs->(($i))."d0";
	if ( $i == nelem($coeffs)-1 ) {
	    print $closing."\n";
	} else {
	    print "+logarithmicMass*(";
	    $closing .= ")";
	}
    }
    &PrettyPlots::Prepare_Dataset(
	\$plot,
	10.0**$fitMass,
	10.0**$fitDistance,
	style      => "line",
	weight     => [5,3],
	color      => $PrettyPlots::colorPairs{${$PrettyPlots::colorPairSequences{'sequence2'}}[$redshiftBin]},
	title      => 'Fit ['.$redshiftLabel.']'
	);
    &PrettyPlots::Prepare_Dataset(
	\$plot,
	10.0**$logarithmicMass,
	10.0**$logarithmicDistance,
	style      => "point",
	weight     => [5,3],
	symbol     => [6,7],
	color      => $PrettyPlots::colorPairs{${$PrettyPlots::colorPairSequences{'sequence1'}}[$redshiftBin]},
	title      => 'VIPERS ['.$redshiftLabel.']'
	);
}

&PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
close($gnuPlot);
&LaTeX::GnuPlot2PDF($plotFileEPS);

exit;

File: constraints/dataAnalysis/stellarMassFunctions_VIPERS_z0_1/massDistanceRelation.py
#!/usr/bin/env python3
# Determine the relation between stellar mass and limiting distance for the VIPERS stellar mass functions.
# Andrew Benson (04-June-2014)

import os
import sys
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import xml.etree.ElementTree as ET
from astropy.cosmology import LambdaCDM
import astropy.units as u

execPath = os.environ.get('GALACTICUS_EXEC_PATH', '').rstrip('/') + '/'
dataPath = os.environ.get('GALACTICUS_DATA_PATH', '').rstrip('/') + '/'

# Define a working directory.
workDirectory = execPath + 'constraints/dataAnalysis/stellarMassFunctions_VIPERS_z0_1/'

# Define survey solid angle (computed from mangle polygons).
solidAngle = 0.003137

# Read the stellar mass function.
xmlFile = dataPath + 'static/observations/massFunctionsStellar/Stellar_Mass_Functions_VIPERS_2013.xml'
tree = ET.parse(xmlFile)
root = tree.getroot()

cosmoElem    = root.find('cosmology')
omegaMatter  = float(cosmoElem.find('omegaMatter'    ).text)
omegaDE      = float(cosmoElem.find('omegaDarkEnergy').text)
hubble       = float(cosmoElem.find('hubble'         ).text)

cosmologyObserved = LambdaCDM(H0=hubble, Om0=omegaMatter, Ode0=omegaDE)

# Estimated sampling rates for each redshift.
samplingRate = np.array([0.44, 0.44, 0.4])

# Create a plot.
fig, ax = plt.subplots(figsize=(8, 6))
ax.set_title('Limiting Distance vs. Stellar Mass for VIPERS Survey')
ax.set_xlabel(r'Stellar mass; $M_\star\;[M_\odot]$')
ax.set_ylabel(r'Limiting distance; $D_{\rm max}$ [Mpc]')
ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlim(3.0e9, 3.0e12)
ax.set_ylim(2000.0, 4000.0)

colors1 = ['steelblue', 'darkorange', 'green']
colors2 = ['navy',      'saddlebrown', 'darkgreen']

massFunction = root.find('massFunction')
columns = massFunction.findall('columns')
for iMF, column in enumerate(columns):
    redshiftBin = iMF

    logarithmicMass = np.array([float(d.text) for d in column.find('mass'        ).findall('datum')])
    massFunction    = np.array([float(d.text) for d in column.find('massFunction').findall('datum')])
    number          = np.array([float(d.text) for d in column.find('number'      ).findall('datum')])
    binWidth        = np.full(len(massFunction), 0.2)
    redshiftMinimum = float(column.find('redshiftLow' ).text)
    redshiftMaximum = float(column.find('redshiftHigh').text)
    redshiftLabel   = f'${redshiftMinimum:.1f}<z<{redshiftMaximum:.1f}$'

    # Find distance to minimum and maximum redshifts.
    distanceMinimum = cosmologyObserved.comoving_distance(redshiftMinimum).to(u.Mpc).value
    distanceMaximum = cosmologyObserved.comoving_distance(redshiftMaximum).to(u.Mpc).value

    # Compute the effective volume.
    volume = number / samplingRate[redshiftBin] / massFunction / binWidth

    # Find the maximum distance.
    distance            = (3.0 * volume / solidAngle + distanceMinimum**3) ** (1.0 / 3.0)
    logarithmicDistance = np.log10(distance)

    # Fit a polynomial to the results.
    fitTo = np.where(
        (massFunction > 0.0) &
        (distance < 0.98 * distanceMaximum) &
        (logarithmicMass < 12)
    )[0]
    coeffsDescending = np.polyfit(logarithmicMass[fitTo], logarithmicDistance[fitTo], 2)
    coeffsAscending  = coeffsDescending[::-1]

    fitMass = np.linspace(4.0, 13.0, 1000)
    fitDistance = np.zeros(len(fitMass))
    closing = ''
    print(f'{redshiftMinimum} < z < {redshiftMaximum}: ', end='')
    for i, c in enumerate(coeffsAscending):
        fitDistance += c * fitMass ** i
        print(f'{c:.10e}d0', end='')
        if i == len(coeffsAscending) - 1:
            print(closing)
        else:
            print('+logarithmicMass*(', end='')
            closing += ')'

    ax.plot(10.0 ** fitMass, 10.0 ** fitDistance,
            color=colors2[redshiftBin], linewidth=2.5, label=f'Fit [{redshiftLabel}]')
    ax.plot(10.0 ** logarithmicMass, 10.0 ** logarithmicDistance,
            'o', color=colors1[redshiftBin], markersize=6, label=f'VIPERS [{redshiftLabel}]')

ax.legend(loc='upper left')
plt.tight_layout()
plotFile = workDirectory + 'massDistanceRelation.pdf'
plt.savefig(plotFile)
plt.close()
